nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/S3EncryptionService.java

```diff
@@ -0,0 +1,189 @@
+
+    public static final PropertyDescriptor ENCRYPTION_STRATEGY = new PropertyDescriptor.Builder()
+            .name("Encryption Strategy")
+            .displayName("Encryption Strategy")
+            .description("Strategy to use for S3 data encryption and decryption.")
+            .allowableValues(NONE, SSE_S3, SSE_KMS, SSE_C, CSE_KMS, CSE_CMK)
+            .required(true)
+            .defaultValue(NONE.getValue())
+            .build();
+
+    public static final PropertyDescriptor ENCRYPTION_VALUE = new PropertyDescriptor.Builder()
+            .name("Key ID or Key Material")
+            .displayName("Key ID or Key Material")
+            .description("Key ID or Key Material used to encrypt and decrypt S3 data.")
+            .required(false)
+            .sensitive(true)
+            .addValidator(new StandardValidators.StringLengthValidator(0, 4096))
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .build();
+
+    public static final PropertyDescriptor REGION = new PropertyDescriptor.Builder()
+            .name("Region")
+            .required(false)
+            .allowableValues(AbstractS3Processor.getAvailableRegions())
+            .defaultValue(AbstractS3Processor.createAllowableValue(Regions.DEFAULT_REGION).getValue())
+            .build();
+
+    private String keyValue = "";
+    private String region = "";
+    private S3EncryptionStrategy encryptionStrategy = new NoOpEncryptionStrategy();
+    private String strategyName = STRATEGY_NAME_NONE;
+
+    @OnEnabled
+    public void onConfigured(final ConfigurationContext context) throws InitializationException {
+        final String newStrategyName = context.getProperty(ENCRYPTION_STRATEGY).getValue();
+        final String newKeyValue = context.getProperty(ENCRYPTION_VALUE).getValue();
+        final S3EncryptionStrategy newEncryptionStrategy = namedStrategies.get(newStrategyName);
+        String newRegion = null;
+
+        if (context.getProperty(REGION) != null ) {
+            newRegion = context.getProperty(REGION).getValue();
+        }
+
+        if (newEncryptionStrategy == null) {
+            final String msg = "No encryption strategy found for name: " + strategyName;
+            logger.warn(msg);
+            throw new InitializationException(msg);
+        }
+
+        strategyName = newStrategyName;
+        encryptionStrategy = newEncryptionStrategy;
+        keyValue = newKeyValue;
+        region = newRegion;
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(final ValidationContext validationContext) {
+        Collection<ValidationResult> validationResults = new ArrayList<>();
+        validationResults.add(encryptionStrategy.validateKey(validationContext.getProperty(ENCRYPTION_VALUE).getValue()));
```

Comment:
Add dependent validation for key ID or material value depending on encryption strategy (i.e. if CMK, when the Base64-encoded key material is entered, decode it and verify it is 16, 24, 32 bytes).

Reply:
I believe this is already done by the custom property validation.